### ISSUE TITLE
[FATAL ../src/RecordSession.cc:1934:process_syscall_entry()]

### ISSUE THREAD

**Issue by tysmith** [rr-debugger/rr#3424]
I hit this while running Firefox. 

```rr record --chaos --disable-cpuid-features-ext 0xdc230000,0x2c42,0xc /home/user/workspace/browsers/m-c-20230109162059-fuzzing-asan-noopt/firefox -no-remote```

I was running rr commit e77b5f8. I only saw it once and could not reproduce. 

This not blocking me, I just thought I should share the log. Feel free to close if this is not actionable.

```
[FATAL ../src/RecordSession.cc:1934:process_syscall_entry()]
 (task 2094714 (rec:2094714) at time 506051)
 -> Assertion `t->desched_rec() || is_rrcall_notify_syscall_hook_exit_syscall( t->regs().original_syscallno(), t->arch()) || t->ip() == t->vm() ->privileged_traced_syscall_ip() .increment_by_syscall_insn_length(t->arch())' failed to hold. Stashed signal pending on syscall entry when it shouldn't be: {signo:SIGSTKFLT,errno:SUCCESS,code:sicode(1)}; regs={ ip:0x56360a3cda71 args:(0,0x1000,0x3,0x22,0xffffffffffffffff,0) orig_syscall: 9 syscallno: -38 }; last_execution_resume=0x7fba5caae5c9; sig ip=0x7fba5caae5c9
Tail of trace dump:
{
  real_time:8231468.702170 global_time:506031, event:`SYSCALL: sched_yield' (state:ENTERING_SYSCALL) tid:2094592, ticks:503189197
rax:0xffffffffffffffda rbx:0x5f18729efde0 rcx:0xffffffffffffffff rdx:0xffffe08e47fc4027 rsi:0x320021ba7e9261 rdi:0x56049df78400 rbp:0x5f18729eff10 rsp:0x5f18729efd58 r8:0x8 r9:0x2 r10:0x17 r11:0x246 r12:0x6400580e7ae0 r13:0x5f18729f1830 r14:0x5f18729f18c0 r15:0x602000830830 rip:0x67ce64853867 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x18 fs_base:0x5f18729f8700 gs_base:0x0
}
{
  real_time:8231468.702307 global_time:506032, event:`SCHED' tid:2094477, ticks:11075078207
rax:0x31565aac9b60 rbx:0x60b00026e120 rcx:0x0 rdx:0xd rsi:0xfd rdi:0xc1680045c26 rbp:0x0 rsp:0x7ffcb48c5678 r8:0x56049de28ef4 r9:0x2 r10:0x1 r11:0x0 r12:0x7fab0fa6f000 r13:0x3 r14:0x56049d56fc08 r15:0x7ffcb48c56d8 rip:0x31565aac9b68 eflags:0x206 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffff09 fs_base:0x67ce6487b4c0 gs_base:0x0
}
{
  real_time:8231468.702313 global_time:506033, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:2094538, ticks:260021142
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x80 rdi:0x60f0001750ec rbp:0x60f0001750c0 rsp:0x7f5aacc94d40 r8:0x0 r9:0x4 r10:0x0 r11:0x246 r12:0x7f5aacc94fa0 r13:0x7f5aacc94fa0 r14:0x4bc20c5af340 r15:0x60f0001750ec rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x4bc20c5b1700 gs_base:0x0
}
{
  real_time:8231468.702556 global_time:506034, event:`SCHED' tid:2094538, ticks:260021235
rax:0x4bc20c5af5c0 rbx:0x4bc20c5af5e0 rcx:0x4bc20c5af420 rdx:0x978c18ade84 rsi:0x1 rdi:0x4bc20c5af5c0 rbp:0x4bc20c5af4f0 rsp:0x4bc20c5af4e0 r8:0x2bb10e051140 r9:0x6160001f2390 r10:0x0 r11:0x0 r12:0x7ffcb48c30ee r13:0x7ffcb48c30ef r14:0x7ffcb48c39e0 r15:0x4bc20c5b0f40 rip:0x64001c640490 eflags:0x206 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffff09 fs_base:0x4bc20c5b1700 gs_base:0x0
}
{
  real_time:8231468.702570 global_time:506035, event:`SYSCALL: nanosleep' (state:EXITING_SYSCALL) tid:2094582, ticks:57399
rax:0x0 rbx:0xcccccccccccccccd rcx:0xffffffffffffffff rdx:0x0 rsi:0x65142b8b7df8 rdi:0x65142b8b7df8 rbp:0x3000 rsp:0x65142b8b7e08 r8:0x0 r9:0x1b0 r10:0x1b0 r11:0x246 r12:0x5556475ed488 r13:0x0 r14:0x0 r15:0x3c1 rip:0x555647251d0e eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x23 fs_base:0x65142b8b8700 gs_base:0x0
}
{
  real_time:8231468.702630 global_time:506036, event:`SYSCALL: openat' (state:ENTERING_SYSCALL) tid:2094582, ticks:57401
rax:0xffffffffffffffda rbx:0x5556470f507b rcx:0xffffffffffffffff rdx:0x0 rsi:0x5556470f507b rdi:0xffffffffffffff9c rbp:0x0 rsp:0x65142b8b7d28 r8:0x0 r9:0x1b0 r10:0x1b0 r11:0x246 r12:0x5556475ed488 r13:0x0 r14:0x0 r15:0x3c1 rip:0x555647251b20 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x101 fs_base:0x65142b8b8700 gs_base:0x0
}
{
  real_time:8231468.702837 global_time:506037, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:2094683, ticks:524810
rax:0xffffffffffffff92 rbx:0xca rcx:0xffffffffffffffff rdx:0x0 rsi:0x89 rdi:0x6100000011dc rbp:0x6100000011b0 rsp:0x7ff5bc4abd40 r8:0x0 r9:0xffffffff r10:0x6ba36e069ce0 r11:0x246 r12:0x7ff5bc4abfa0 r13:0x7ff5bc4abfa0 r14:0x6100000011dc r15:0x6ba36e069b90 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x6ba36e06c700 gs_base:0x0
}
{
  real_time:8231468.702957 global_time:506038, event:`SYSCALLBUF_FLUSH' tid:2094683, ticks:526193
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
}
{
  real_time:8231468.702960 global_time:506039, event:`SYSCALL: futex' (state:ENTERING_SYSCALL) tid:2094683, ticks:526193
rax:0xffffffffffffffda rbx:0xca rcx:0xffffffffffffffff rdx:0x1 rsi:0x81 rdi:0x610000001180 rbp:0x6ba36e069da0 rsp:0x7ff5bc4abd40 r8:0x0 r9:0x2 r10:0x0 r11:0x246 r12:0x7ff5bc4abfa0 r13:0x7ff5bc4abfa0 r14:0x7ffeb2727f40 r15:0x6ba36e06bf40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x6ba36e06c700 gs_base:0x0
}
{
  real_time:8231468.702962 global_time:506040, event:`SYSCALLBUF_RESET' tid:2094683, ticks:526193
}
{
  real_time:8231468.702994 global_time:506041, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:2094683, ticks:526193
rax:0x0 rbx:0xca rcx:0xffffffffffffffff rdx:0x1 rsi:0x81 rdi:0x610000001180 rbp:0x6ba36e069da0 rsp:0x7ff5bc4abd40 r8:0x0 r9:0x2 r10:0x0 r11:0x246 r12:0x7ff5bc4abfa0 r13:0x7ff5bc4abfa0 r14:0x7ffeb2727f40 r15:0x6ba36e06bf40 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x6ba36e06c700 gs_base:0x0
}
{
  real_time:8231468.703064 global_time:506042, event:`SYSCALLBUF_FLUSH' tid:2094683, ticks:527317
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
}
{
  real_time:8231468.703067 global_time:506043, event:`SYSCALL: write' (state:ENTERING_SYSCALL) tid:2094683, ticks:527317
rax:0xffffffffffffffda rbx:0x1 rcx:0xffffffffffffffff rdx:0x1 rsi:0x6ba36e0686c0 rdi:0x12 rbp:0x6ba36e068690 rsp:0x7ff5bc4abd40 r8:0x0 r9:0x1 r10:0x0 r11:0x246 r12:0x7ff5bc4abfa0 r13:0x7ff5bc4abfa0 r14:0x1 r15:0x6ba36e0686c0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x1 fs_base:0x6ba36e06c700 gs_base:0x0
}
{
  real_time:8231468.703069 global_time:506044, event:`SYSCALLBUF_RESET' tid:2094683, ticks:527317
}
{
  real_time:8231468.703213 global_time:506045, event:`SYSCALL: futex' (state:ENTERING_SYSCALL) tid:2094536, ticks:29664675
rax:0xffffffffffffffda rbx:0xca rcx:0xffffffffffffffff rdx:0x1 rsi:0x81 rdi:0x60f0001751d8 rbp:0x81 rsp:0x7f5aad29cd40 r8:0x60f0001751b0 r9:0x60f0001751d0 r10:0x0 r11:0x246 r12:0x7f5aad29cfa0 r13:0x7f5aad29cfa0 r14:0xca r15:0x60f0001751d8 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x15c82e103700 gs_base:0x0
}
{
  real_time:8231468.703251 global_time:506046, event:`SYSCALL: futex' (state:EXITING_SYSCALL) tid:2094536, ticks:29664675
rax:0x1 rbx:0xca rcx:0xffffffffffffffff rdx:0x1 rsi:0x81 rdi:0x60f0001751d8 rbp:0x81 rsp:0x7f5aad29cd40 r8:0x60f0001751b0 r9:0x60f0001751d0 r10:0x0 r11:0x246 r12:0x7f5aad29cfa0 r13:0x7f5aad29cfa0 r14:0xca r15:0x60f0001751d8 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xca fs_base:0x15c82e103700 gs_base:0x0
}
{
  real_time:8231468.703326 global_time:506047, event:`SYSCALL: openat' (state:ENTERING_SYSCALL) tid:2094557, ticks:65964
rax:0xffffffffffffffda rbx:0x55604f49f07b rcx:0xffffffffffffffff rdx:0x0 rsi:0x55604f49f07b rdi:0xffffffffffffff9c rbp:0x0 rsp:0x3175687e2d28 r8:0x0 r9:0x1b0 r10:0x1b0 r11:0x246 r12:0x55604f997488 r13:0x0 r14:0x0 r15:0x24d rip:0x55604f5fbb20 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x101 fs_base:0x3175687e3700 gs_base:0x0
}
{
  real_time:8231468.703442 global_time:506048, event:`SYSCALLBUF_FLUSH' tid:2094578, ticks:1487591225
  { syscall:'clock_gettime', ret:0x0, size:0x20 }
}
{
  real_time:8231468.703445 global_time:506049, event:`SYSCALL: write' (state:ENTERING_SYSCALL) tid:2094578, ticks:1487591225
rax:0xffffffffffffffda rbx:0x1 rcx:0xffffffffffffffff rdx:0x1 rsi:0x7ffc2ec82020 rdi:0x18 rbp:0x7ffc2ec81ff0 rsp:0x681ffd40 r8:0x0 r9:0x1 r10:0x0 r11:0x246 r12:0x681fffa0 r13:0x681fffa0 r14:0x1 r15:0x7ffc2ec82020 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x1 fs_base:0x70af579df4c0 gs_base:0x0
}
{
  real_time:8231468.703448 global_time:506050, event:`SYSCALLBUF_RESET' tid:2094578, ticks:1487591225
}
[FATAL ../src/log.cc:444:emergency_debug()] (session doesn't look interactive, aborting emergency debugging)
=== Start rr backtrace:
rr(_ZN2rr15notifying_abortEv+0xe0)[0x5db6e0]
rr(_ZN2rr12FatalOstreamD1Ev+0x67)[0x4a8297]
rr(_ZN2rr21EmergencyDebugOstreamD1Ev+0x29e)[0x4a89de]
rr(_ZN2rr13RecordSession21process_syscall_entryEPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultENS_13SupportedArchE+0x3ec)[0x4d131c]
rr(_ZN2rr13RecordSession29handle_seccomp_traced_syscallEPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultEPb+0x684)[0x4d0a14]
rr(_ZN2rr13RecordSession19handle_ptrace_eventEPPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultEPb+0x5a2)[0x4d2e62]
rr(_ZN2rr13RecordSession11record_stepEv+0x368)[0x4de978]
rr(_ZN2rr13RecordCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE+0x1a7e)[0x4cf4ce]
rr(main+0x39b)[0x5e5dcb]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7fbdadcae083]
rr(_start+0x2e)[0x42cece]
=== End rr backtrace
```

**Comment by rocallahan** [comment]
rr should have processed the signal instead of issuing the PTRACE_CONT that led to the mmap syscall. It's hard to determine what happened instead.

**Comment by markus456** [comment]
I seem to be hitting a similar assertion on Fedora 38 with rr 5.7.0:
```
[FATAL src/RecordSession.cc:1944:process_syscall_entry()] 
 (task 42105 (rec:42105) at time 1951)
 -> Assertion `t->desched_rec() || is_rrcall_notify_syscall_hook_exit_syscall( t->regs().original_syscallno(), t->arch()) || t->ip() == t->vm() ->privileged_traced_syscall_ip() .increment_by_syscall_insn_length(t->arch())' failed to hold. Stashed signal pending on syscall entry when it shouldn't be: {signo:SIGSTOP,errno:SUCCESS,code:SI_TKILL}; regs={ ip:0x7f203499577b args:(0xefffffffffffffff,0,0,0,0,0) orig_syscall: 24 syscallno: -38 }; last_execution_resume=0x7f203499577b; sig ip=0x7f203499577b
```
I was running a program compiled with both ASAN and UBSAN and the leak check appears to be the one that's causing it. Disabling it with `ASAN_OPTIONS=detect_leaks=0` makes it go away. Interestingly enough, the program I was running  did cause ASAN to detect a problem and exit but even if I use `ASAN_OPTIONS=detect_leaks=0,abort_on_error=1`, `rr` manages to complete successfully.  Maybe something in the LSAN leak check is causing it?

Here's the stacktraces from rr and gdb:

```=== Start rr backtrace:
rr(_ZN2rr13dump_rr_stackEv+0x5e)[0x562b4bd7b52e]
rr(_ZN2rr9GdbServer15emergency_debugEPNS_4TaskE+0x15e)[0x562b4bc53eae]
rr(+0xc5236)[0x562b4bc66236]
rr(+0xc5555)[0x562b4bc66555]
rr(_ZN2rr13RecordSession21process_syscall_entryEPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultENS_13SupportedArchE+0x346)[0x562b4bc958f6]
rr(_ZN2rr13RecordSession29handle_seccomp_traced_syscallEPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultEPb+0x36d)[0x562b4bc9037d]
rr(_ZN2rr13RecordSession19handle_ptrace_eventEPPNS_10RecordTaskEPNS0_9StepStateEPNS0_12RecordResultEPb+0x747)[0x562b4bc91127]
rr(_ZN2rr13RecordSession11record_stepEv+0x2a0)[0x562b4bc96490]
rr(_ZN2rr13RecordCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE+0x861)[0x562b4bc8a611]
rr(main+0x166)[0x562b4bbed036]
/lib64/libc.so.6(+0x27b8a)[0x7f7b17168b8a]
/lib64/libc.so.6(__libc_start_main+0x8b)[0x7f7b17168c4b]
rr(_start+0x25)[0x562b4bbee855]
=== End rr backtrace



Thread 1 (Thread 42105.42105 (compare_canonic)):
#0  0x00007f203499577b in sched_yield () at /builddir/build/BUILD/rr-5.7.0/src/preload/overrides.c:247
#1  0x00007f2034301d75 in __sanitizer::StopTheWorld(void (*)(__sanitizer::SuspendedThreadsList const&, void*), void*) () from /lib64/libasan.so.8
#2  0x00007f203431047e in __lsan::LockStuffAndStopTheWorldCallback(dl_phdr_info*, unsigned long, void*) () from /lib64/libasan.so.8
#3  0x00007f20335795f2 in dl_iterate_phdr () from /lib64/libc.so.6
#4  0x00007f20343107e0 in __lsan::LockStuffAndStopTheWorld(void (*)(__sanitizer::SuspendedThreadsList const&, void*), __lsan::CheckForLeaksParam*) () from /lib64/libasan.so.8
#5  0x00007f203430e970 in __lsan::CheckForLeaks() () from /lib64/libasan.so.8
#6  0x00007f203430ebb8 in __lsan::DoLeakCheck() () from /lib64/libasan.so.8
#7  0x00007f2033461c3d in __cxa_finalize () from /lib64/libc.so.6
#8  0x00007f2034226927 in __do_global_dtors_aux () from /lib64/libasan.so.8
#9  0x00007fff4879abd0 in ?? ()
#10 0x00007f20349bf0f2 in _dl_call_fini (closure_map=0x7f20349a4000) at dl-call_fini.c:43
Backtrace stopped: frame did not save the PC
```

**Comment by rocallahan** [comment]
@markus456 so this is reproducible for you? If it is, maybe it's worth trying to come up with a small testcase that we can work with.

**Comment by markus456** [comment]
Here's a small program that seems to reproduce it:

```
#include <iostream>
#include <cstring>

class Thing
{
public:
    Thing()
    {
        buffer = new char[128];
    }

    ~Thing()
    {
        delete[] buffer;
    }

    char* buffer;
};

Thing thing;

int main(int argc, char** argv)
{
    std::string line;
    std::cin.read(thing.buffer, 128);
    thing.buffer[std::cin.gcount()] = 0;
    std::cout << thing.buffer << std::endl;
    return 0;
}
```

I compiled it with `g++ test.cc -fsanitize=address -fsanitize=undefined -g` with GCC 13.2.1.

Here's how I run it:

```
rr ./a.out <<< "helloooooooooooooooooooooooooooooo"
```

Here's the output: [output.txt](https://github.com/rr-debugger/rr/files/14006160/output.txt)

**Comment by rocallahan** [comment]
Those steps doesn't reproduce it for me with rr `master`. I'm even using Fedora 38 and its gcc 13.2.1.

**Comment by markus456** [comment]
OK, I can confirm that it works correctly when I build it from `master`.

**Comment by rocallahan** [comment]
Thanks!

**Comment by KJTsanaktsidis** [comment]
I had this happen on master just now (built from https://github.com/rr-debugger/rr/commit/a2b50f4945fb10598d6829e10f192c983359b5b6)

Crashlog: https://gist.github.com/KJTsanaktsidis/e247374c08d495d8765d4933f72bb7b3.

This is from running the Ruby test suite through RR. I'm working on trying to come up with a useful reproduction of it (it's happening only occasionally.

One thing to note - I'm NOT using a distribution provided package, and I AM using the very latest hot-off-the-press copy of rr, but I AM building rr into an RPM package which means it IS using the soup of distribution-provided CFLAGS. I wonder if there's something wrong in those?

The rr package I'm using: https://pagure.io/asan-dist/rr-asan-dist/blob/asan-dist/rawhide/f/rr.spec (built: https://download.copr.fedorainfracloud.org/results/kjtsanaktsidis/asan-dist/fedora-40-x86_64/07690179-rr-asan-dist/rr-asan-dist-5.8.0-20240629.1.fc40.x86_64.rpm)

**Comment by KJTsanaktsidis** [comment]
OK, thanks to @markus456's reproduction I can definitely confirm its a probem with LTO enabled. I'm going to go open another issue for it.
